fabfile.py

```diff
@@ -1,197 +1,518 @@
+
+VIRTUAL_ENV_PATH = "/home/{usr}/.virtualenvs/{release_name}"
+PROJECT_ROOT = "/usr/lib/{unix_user}".format(unix_user=UNIX_USER)
+PROJECT_DIRECTORY = "{project_root}/{release_name}"
+BACKUP_DIR = "{project_root}/var".format(project_root=PROJECT_ROOT)
+GIT_URL = "https://github.com/openhealthcare/elcid-rfh"
+
+# the daily back up
+BACKUP_NAME = "{backup_dir}/back.{dt}.{db_name}.sql"
+
+# the release back up is take just before the release, and then restored
+RELEASE_BACKUP_NAME = "{backup_dir}/release.{dt}.{db_name}.sql"
+
+MODULE_NAME = "elcid"
+PROJECT_NAME = MODULE_NAME
+CRON_COMMENT = "{} back up".format(MODULE_NAME)
+DB_COMMAND_PREFIX = "sudo -u postgres psql --command"
+TEMPLATE_DIR = os.path.abspath(os.path.dirname(__file__))
+PRIVATE_SETTINGS = "{project_root}/private_settings.json".format(
+    project_root=PROJECT_ROOT
+)
+jinja_env = Environment(loader=FileSystemLoader(TEMPLATE_DIR))
+
+
+class Env(object):
+    def __init__(self, branch):
+        self.branch = branch
+
+    @property
+    def project_directory(self):
+        return PROJECT_DIRECTORY.format(
+            project_root=PROJECT_ROOT,
+            release_name=self.release_name,
+        )
 
+    @property
+    def release_name(self):
+        return RELEASE_NAME.format(branch=self.branch)
 
-env.hosts = ["elcid-rfh-test.openhealthcare.org.uk"]
-env.user = "ubuntu"
-virtual_env_name = "elcid-rfh"
-fabfile_dir = os.path.realpath(os.path.dirname(__file__))
+    @property
+    def virtual_env_path(self):
+        return VIRTUAL_ENV_PATH.format(
+            usr=UNIX_USER,
+            release_name=self.release_name
+        )
 
+    @property
+    def database_name(self):
+        return self.release_name.replace("-", "_").replace(".", "")
+
+    @property
+    def backup_name(self):
+        now = datetime.datetime.now()
+        return BACKUP_NAME.format(
+            backup_dir=BACKUP_DIR,
+            dt=now.strftime("%d.%m.%Y"),
+            db_name=self.database_name
+        )
 
-def check_for_uncommitted():
-    changes = local("git status --porcelain", capture=True)
-    return len(changes)
+    @property
+    def release_backup_name(self):
+        return RELEASE_BACKUP_NAME.format(
+            backup_dir=BACKUP_DIR,
+            dt=datetime.datetime.now().strftime("%d.%m.%Y.%H.%M"),
+            db_name=self.database_name
+        )
 
 
-def get_requirements():
-    """
-    looks for a requirements file in the same directory as the
-    fabfile. Parses it,
-    """
+def run_management_command(some_command, env):
+    with lcd(env.project_directory):
+        cmd = "{0}/bin/python manage.py {1}".format(
+            env.virtual_env_path, some_command
+        )
+        result = local(cmd, capture=True)
+    return result
 
-    with lcd(fabfile_dir):
-        requirements = local("less requirements.txt", capture=True).split("\n")
 
-        package_to_version = {}
+def pip_create_virtual_env(new_env):
+    local("rm -rf {}".format(new_env.virtual_env_path))
```

Comment:
We definitely want to always trash an existing virtualenv if we re-deploy a version?
What if it's during a deploy_prod call ?